crates/walrus-service/src/node/blob_sync.rs

```diff
@@ -333,16 +336,19 @@ impl BlobSyncHandler {
                 entry.insert(InProgressSyncHandle {
                     cancel_token,
                     blob_sync_handle: Some(sync_handle),
+                    finish_notify: finish_notify.clone(),
                 });
+                finish_notify
             }
-            Entry::Occupied(_) => {
+            Entry::Occupied(existing_sync) => {
                 // A blob sync with a lower sequence number is already in progress. We can safely
                 // try to increase the event cursor since it will only be advanced once that sync is
                 // finished or cancelled due to an invalid blob event.
                 event_handle.mark_as_complete();
-                finish_notify.notify_one();
+
+                existing_sync.get().finish_notify.clone()
             }
-        }
+        };
         Ok(finish_notify)
```

Comment:
Before, when there an existing blob sync task, start_sync returns a finish_notify that is notified immediately. Now, we return a notify that attaches to the existing blob sync tasks.

Reply:
Remark/question: I think this makes sense to me. But is there a concrete need to make this change?

Reply:
Yes, the issue without this change is that, say RecoveryInProgress(epoch=9) is happening, and started all the blob syncs, and then we transition into epoch 10. All the blob syncs initiated prior are still running. In the node recovery, we loop through all the certified blobs, find ones that needs recovery, and start the recovery. Without this change, the task handle returned by `start_sync` will finish immediately. This causes the loop to finish very quickly. However, when the loop runs again to see we indeed finished recovering all the blobs, they still see blobs that need recover. So we ended up running the recovery loop over and over. The motivation for this change is the observed behavior in simtest.